src/algorithms/highlevel/crosssimilaritymatrix.cpp

```diff
@@ -0,0 +1,233 @@
+
+
+    // Construct thresholded similarity matrix on axis X
+    for (size_t k=0; k<pdistances.size(); k++) {
+        for (size_t l=0; l<pdistances[k].size(); l++) {
+            similarityX[k][l] = ephX[k][0] - pdistances[k][l];
+        }
+    }
+
+    // Construct thresholded similarity matrix on axis Y
+    for (size_t u=0; u<tpDistances.size(); u++) {
+        for (size_t v=0; v<tpDistances[u].size(); v++) {
+            similarityY[u][v] = ephY[v][0] - tpDistances[u][v];
+        }
+    }
+
+    /*
+    if (similarityX.empty() == 0) {
+      throw EssentiaException("Sim X is empty");
+    }
+
+    if (similarityY.empty() == 0) {
+      throw EssentiaException("Sim Y is empty");
+    }
+    */
+    // Clear memory
+    std::vector<std::vector<Real> >().swap(ephX);
+    std::vector<std::vector<Real> >().swap(ephY);
+
+    // Binarise the array with heavisideStepFunction
+    heavisideStepFunction(similarityX);
+    heavisideStepFunction(similarityY);
+
+    csm.assign(similarityX.size(), std::vector<Real>(similarityY.size(), 0));
+    // tranpose similarityY vector for dot product
+    std::vector<std::vector<Real> > tSimilarityY = transpose(similarityY);
+    // clear memory
+    std::vector<std::vector<Real> >().swap(similarityY);
+
+    // Finally we construct out cross similarity matrix by doing dot product
+    for (size_t x=0; x<similarityX.size(); x++) {
+        for (size_t y=0; y<tSimilarityY.size(); y++) {
+            csm[x][y] = dotProduct(similarityX[x], tSimilarityY[y]);
+        }
+    }
+}
+
+
+// Construct a stacked chroma embedding from an input chroma audio feature vector [TODO] future use beat-synchronised stacked embeddings
+std::vector<std::vector<Real> > CrossSimilarityMatrix::toTimeEmbedding(std::vector<std::vector<Real> >& inputArray, int m, int tau) const {
+
+    int stopIdx;
+    int increment = m*tau;
+    int frameSize = inputArray[0].size() - increment;
+    int yDim = inputArray.size() * m;
+    std::vector<std::vector<Real> > timeEmbedding(frameSize, std::vector<Real>(yDim, 0));
+    std::vector<Real> tempRow;
+    tempRow.reserve(yDim*tau);
+
+    for (int i=0; i<frameSize; i+=tau) {
+        stopIdx = i + increment;
+        for (int startTime=i; startTime<=stopIdx; startTime+=tau) {
+            if (startTime == i) {
+                tempRow = inputArray[startTime];
+            }
+            else {
+                tempRow.insert(tempRow.end(), inputArray[startTime].begin(), inputArray[startTime].end());
+            }
+        timeEmbedding[i] = tempRow;
+        };
+    }
+    return timeEmbedding;
+}
+
+
+// computes global averaged chroma hpcp as described in [1]
+std::vector<Real> CrossSimilarityMatrix::globalAverageChroma(std::vector<std::vector<Real> >& inputFeature) const {
+
+    std::vector<Real> globalChroma;
+    std::vector<std::vector<Real> > featureCopy;
+    // tranpose the array from (time_axis, chroma_bin) to (chroma_bin, time-axis)for easy calculation
```

Comment:
To avoid memory allocation for the copy, we can compute the sum value directly on the (time_axis, chroma_bin) matrix.